Change cluster/gce/util to wait for firewall rule creation and fail on error

brianpursley opened the pull request:
#### What type of PR is this?
/kind flake

#### What this PR does / why we need it:
I suspect that some e2e flakes are actually cluster setup failures that go unnoticed.  

This PR attempts to improve e2e test flakiness by waiting for firewall rule creation background commands to complete and report any errors that occur, immediately, instead of waiting for an e2e test to fail later on.

#### Which issue(s) this PR fixes:
Fixes #111580

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

brianpursley commented:
/retest

bowei commented:
Any historical info you can dig up why those are run asynchronously (change looks ok to me generally)

brianpursley commented:
> Any historical info you can dig up why those are run asynchronously (change looks ok to me generally)

I don't know why, but it does give me pause. Perhaps creating a firewall rule was slow, but I would only be guessing as to the reason.

This code looks to have been added in 2016 (https://github.com/kubernetes/kubernetes/commit/b8cbac6753e9990e75e7fcad8c347faa98846707) by @jszczepkowski. It looks like maybe he hasn't been active on github lately, but I'm curious to know if he has any insight as to the reason to background the firewall creation calls.

brianpursley commented:
I posted this on the sig-testing slack channel, but here is one occurrence where this happened:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/111577/pull-kubernetes-e2e-gce-ubuntu-containerd/1553389589622689792

Then in the build log, you can see this happened earlier, when util.sh was running, but it didn't fail:
```
W0730 15:08:03.961] Creating firewall...
W0730 15:08:03.961] failed.
W0730 15:08:03.961] ERROR: (gcloud.compute.firewall-rules.create) Could not fetch resource:
W0730 15:08:03.961]  - Internal error. Please try again or contact Google Support. (Code: '5E50721D74B2E.6B58410.2708EFD8')
W0730 15:08:03.961] 
```

I suspect this might be the problem for all of these flakes:
https://storage.googleapis.com/k8s-triage/index.html?pr=1&test=Firewall.rule%20should%20have%20correct%20firewall%20rules%20for%20e2e%20cluster#55aa33a51fc2545fbb0b
![image](https://user-images.githubusercontent.com/4977542/182656102-02a29fba-17fa-4515-bf52-ab9d17ccdd0a.png)

This PR would not fix the problem though, just make it more obvious where it is occurring, and avoid needless test running when it is doomed to fail.

Probably there needs to be some investigation why the firewall rule creation is actually failing, and fix that, or maybe retry if it is impossible to reliably create firewall rules.

bowei commented:
Ping me on slack when you are done with experimentation -- generally this looks ok.

brianpursley committed:
Change cluster/gce/util to wait for firewall rule creation

Add kube::util::wait-for-jobs in create-network() to wait
for background firewall rule creation, and if any fail, log
a message and exit with the failed job's return code.

brianpursley commented:
I learned a little more today after reading the existing code and I think I might have determine a better way of handing failures without removing the backgrounding of firewall rule creation.

Other async/background commands in `util.sh` are using `kube::util::wait-for-jobs` to wait for their completion before continuing.  I've updated this PR to use `kube::util::wait-for-jobs` and if any of the jobs fail, log a message and return the exit code.

`create-network()` was missing this wait, so I think just by adding that it should improve things.

I also updated the wait in `create-nodes-firewall()` to exit on failure.  

The other ones I left non-exiting because they are teardown-related and should probably not fail the test run. I assume those are correct as-is currently.

cc @BenTheElder since we were discussing this PR on slack

brianpursley commented:
/assign @bowei

bowei commented:
/lgtm
/approve
